driver-core/src/main/com/mongodb/internal/session/ServerSessionPool.java

```diff
@@ -77,57 +72,46 @@ public ServerSessionPool(final Cluster cluster, @Nullable final ServerApi server
 
     public ServerSession get() {
         isTrue("server session pool is open", !closed);
-        ServerSessionImpl serverSession = serverSessionPool.get();
-        while (shouldPrune(serverSession)) {
-            serverSessionPool.release(serverSession, true);
-            serverSession = serverSessionPool.get();
+        ServerSessionImpl serverSession = available.pollLast();
+        while (serverSession != null && shouldPrune(serverSession)) {
+            serverSession.close();
+            serverSession = available.pollLast();
+        }
+        if (serverSession == null) {
+            serverSession = new ServerSessionImpl();
         }
+        inUseCount.incrementAndGet();
         return serverSession;
     }
 
     public void release(final ServerSession serverSession) {
-        serverSessionPool.release((ServerSessionImpl) serverSession);
-        serverSessionPool.prune();
+        inUseCount.decrementAndGet();
+        ServerSessionImpl serverSessionImpl = (ServerSessionImpl) serverSession;
+        if (serverSessionImpl.isMarkedDirty()) {
+            serverSessionImpl.close();
+        } else {
+            available.addLast(serverSessionImpl);
```

Comment:
Does adding to last make it more likely that an earlier session in the deque will be left in a should be pruned state but never pruned?

Reply:
This is not a change in behavior.  ConcurrentPool also polls last and adds to last. 

What makes it more likely that an earlier session will not be pruned  is the fact that we're no longer pruning in #release, but only in #get.  But there is an upper bound on the size of the pool, as @stIncMale explained to me: it's bounded by the total number of available connections (because we only allocate a new session after checking out a connection).  I think this is ok because each item in the pool is quite small, so even if there are thousands of stale ones it's not likely to be noticed.

Reply:
That's fine - so by adding it to the end of the queue you make it more likely a user has a non-stale session. And as soon as they do they will all be cleaned.